File: swift/core/Sources/A2UICore/State/SurfaceComponentsModel.swift
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Combine
import Foundation

/// Manages a flat collection of ``ComponentModel`` instances by ID.
///
/// Mirrors `SurfaceComponentsModel` in the core blueprint and
/// `web_core`. This is a pure data container with no schema awareness
/// or validation logic — the `MessageProcessor` handles validation
/// before adding components here.
public final class SurfaceComponentsModel: @unchecked Sendable, ObservableObject {

  private let lock = NSRecursiveLock()
  public let objectWillChange = ObservableObjectPublisher()

  public private(set) var components: [String: ComponentModel] = [:] {
    willSet {
      objectWillChange.send()
    }
    didSet {
      componentsDidChange.send(components)
    }
  }

  /// Synchronous publisher that emits after components have been modified.
  public let componentsDidChange: CurrentValueSubject<[String: ComponentModel], Never>

  /// Creates an empty components model.
  public init() {
    self.componentsDidChange = CurrentValueSubject([:])
  }

  /// Retrieves the component with the given ID.
  ///
  /// - Parameter id: The component ID to look up.
  /// - Returns: The `ComponentModel` if found, otherwise `nil`.
  public func get(_ id: String) -> ComponentModel? {
    lock.withLock { components[id] }
  }

  /// Adds or replaces a component in the collection.
  ///
  /// - Parameter component: The component model to add.
  public func addComponent(_ component: ComponentModel) {
    lock.withLock {
      components[component.id] = component
    }
  }

  /// Removes the component with the given ID.
  ///
  /// - Parameter id: The component ID to remove.
  public func removeComponent(_ id: String) {
    _ = lock.withLock {
      components.removeValue(forKey: id)
    }
  }
}

File: agent_sdks/python/a2ui_core/src/a2ui/core/state/surface_components_model.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Optional
from ..common.events import EventSource
from .component_model import ComponentModel


class SurfaceComponentsModel:
    """Manages the adjacency map of component configs in a surface."""

    def __init__(self) -> None:
        self._components: Dict[str, ComponentModel] = {}
        self.on_created = EventSource()
        self.on_deleted = EventSource()

    def get(self, component_id: str) -> Optional[ComponentModel]:
        return self._components.get(component_id)

    def get_all(self) -> Dict[str, ComponentModel]:
        return self._components

    def add_component(self, component: ComponentModel) -> None:
        if component.id in self._components:
            raise ValueError(f"Component with id '{component.id}' already exists.")
        self._components[component.id] = component
        self.on_created.emit(component)

    def remove_component(self, component_id: str) -> None:
        if component_id in self._components:
            comp = self._components[component_id]
            del self._components[component_id]
            comp.dispose()
            self.on_deleted.emit(component_id)

    def dispose(self) -> None:
        """Disposes of the model and all its components."""
        for component in list(self._components.values()):
            component.dispose()
        self._components.clear()
        if hasattr(self.on_created, "dispose"):
            self.on_created.dispose()
        if hasattr(self.on_deleted, "dispose"):
            self.on_deleted.dispose()
